kernel/src/mem/phys.rs

```diff
@@ -113,33 +114,39 @@ pub(in crate::mem) fn init_stage2() {
     2. USABLE regions are sorted by base address, lowest to highest
     3. USABLE regions are 4KiB aligned (address and length)
      */
-    let highest_usable_address = {
-        let last_usable_region = regions
-            .iter()
-            .rev()
-            .find(|r| r.entry_type == EntryType::USABLE)
-            .expect("no usable regions");
-        last_usable_region.base + last_usable_region.length
-    };
 
-    let mut frames = vec![FrameState::Unusable; (highest_usable_address / Size4KiB::SIZE) as usize];
-
-    stage1
-        .usable_frames()
-        .take(stage_one_next_free)
-        .for_each(|frame| {
-            let index = frame.start_address().as_u64() / Size4KiB::SIZE;
-            frames[index as usize] = FrameState::Allocated;
-        });
-    stage1
-        .usable_frames()
-        .skip(stage_one_next_free)
-        .for_each(|frame| {
-            let index = frame.start_address().as_u64() / Size4KiB::SIZE;
-            frames[index as usize] = FrameState::Free;
-        });
-
-    let bitmap_allocator = PhysicalMemoryManager::new(frames);
+    // Build memory regions for usable regions
+    // Preallocate to avoid fragmentation in stage1 (which can't deallocate)
+    let usable_region_count = regions
+        .iter()
+        .filter(|r| r.entry_type == EntryType::USABLE)
+        .count();
+    let mut memory_regions = Vec::with_capacity(usable_region_count);
+
+    for entry in regions.iter().filter(|r| r.entry_type == EntryType::USABLE) {
+        let num_frames = (entry.length / Size4KiB::SIZE) as usize;
+        let region = kernel_physical_memory::MemoryRegion::new(
+            entry.base,
+            num_frames,
+            kernel_physical_memory::FrameState::Free,
+        );
+        memory_regions.push(region);
```

Comment:
To make sure that this doesn't allocate, use push_within_capacity. This can panic. We are early enough in the initialization and this is not recoverable and a developer error.